### PR TITLE
add base_namespace config value and use it for Mailable instanciation

### LINKED ISSUE #12: Add a base namespace configuration option
Add a `base_namespace` configuration key (default: `App\Mails`), this would allow

```
php artisan mail:send-test MyMailable
```

instead of

```
php artisan mail:send-test "App\Mails\MyMailable"
```

First the package should check if the mailable class exists as is `MyMailable`, then prepend the base namespace if nothing was found and try again.

### PR DESCRIPTION
**Pull request by fetzi**
fixes #12

### CODE REVIEW

File: src/SendTestMail.php [hunk 1]
```diff
@@ -39,7 +39,7 @@ protected function guardAgainstInvalidArguments()
             throw new InvalidArgumentException("`{$this->argument('recipient')}` is not a valid e-mail address");
         }
 
-        if (! class_exists($this->argument('mailableClass'))) {
+        if (! $this->vaildateMailable($this->argument('mailableClass'))) {
```
Comment (freekmurze) [inline src/SendTestMail.php]: Typo in function name

File: config/mailable-test.php [hunk 2]
```diff
@@ -8,4 +8,9 @@
      * extends \Spatie\MailableTest\ArgumentValueProvider::class
      */
     'argument_value_provider_class' => \Spatie\MailableTest\FakerArgumentValueProvider::class,
+
+    /*
+     * base namespace Mailable classes
```
Comment (freekmurze) [inline config/mailable-test.php]: Make the first character a capital

File: src/MailableFactory.php [hunk 3]
```diff
@@ -20,7 +20,11 @@ public function __construct(ArgumentValueProvider $argumentValueProvider)
     public function getInstance(string $mailableClass, string $toEmail, $defaultValues): Mailable
     {
         if (! class_exists($mailableClass)) {
-            throw MailableDoesntExist::withClass($mailableClass);
+            $mailableClass = sprintf('%s\\%s', config('mailable-test.base_namespace'), $mailableClass);
+
+            if (! class_exists($mailableClass)) {
+                throw MailableDoesntExist::withClass($mailableClass);
```
Comment (freekmurze) [inline src/MailableFactory.php]: Rename to `MailableDoesNotExist`

File: tests/SendTestMailTest.php [hunk 4]
```diff
@@ -79,4 +80,41 @@ public function it_will_throw_an_exception_when_passing_invalid_values()
             '--values' => 'myInteger',
         ]);
     }
+
+    /** @test */
+    public function it_will_use_base_namespace_variable_if_class_is_not_found()
+    {
+        Config::set('mailable-test.base_namespace', 'Spatie\\MailableTest\\Test');
```
Comment (freekmurze) [inline tests/SendTestMailTest.php]: Use `config` function instead of the facade.

Comment (freekmurze) [inline tests/SendTestMailTest.php]: Preferably users shouldn't have to type double slashes in the config file. Could change this line to

```
config('mailable-test.base_namespace', Spatie\MailableTest\Test::class);
```

and make sure the test passes.

File: README.md [hunk 5]
```diff
@@ -53,13 +53,18 @@ This is the contents of the published config file:
...
+     * of a mailable. This can be any class as long as it
      * extends \Spatie\MailableTest\ArgumentValueProvider::class
      */
-    'argument_value_provider_class' => \Spatie\MailableTest\FakeArgumentValueProvider::class,
+    'argument_value_provider_class' => \Spatie\MailableTest\FakerArgumentValueProvider::class,
+
+    /*
+     * base namespace Mailable classes
+     */
+    'base_namespace' => 'App\\Mails',
```
Comment (freekmurze) [inline README.md]: Could you make sure this option will work without users having to type double slashes?

**Review by freekmurze — CHANGES_REQUESTED** [review]
Thank you for this. Could you take care of these comments?

**Comment by fetzi** [comment]
@freekmurze if fixed all comments from your review.

**Review by freekmurze — APPROVED** [review]

**Comment by freekmurze** [comment]
Ok for me, thanks

@sebastiandedeyne will take care of this further.

File: config/mailable-test.php [hunk 6]
```diff
@@ -8,4 +8,9 @@
      * extends \Spatie\MailableTest\ArgumentValueProvider::class
      */
     'argument_value_provider_class' => \Spatie\MailableTest\FakerArgumentValueProvider::class,
+
+    /*
+     * Base namespace Mailable classes
+     */
+    'base_namespace' => 'App\Mails',
```
Comment (sebastiandedeyne) [inline config/mailable-test.php]: Can this be changed to `App\Mail`? (sorry, was wrong in the original issue)

File: src/MailableFactory.php [hunk 7]
```diff
@@ -20,7 +20,11 @@ public function __construct(ArgumentValueProvider $argumentValueProvider)
     public function getInstance(string $mailableClass, string $toEmail, $defaultValues): Mailable
     {
         if (! class_exists($mailableClass)) {
-            throw MailableDoesntExist::withClass($mailableClass);
+            $mailableClass = sprintf('%s\\%s', config('mailable-test.base_namespace'), $mailableClass);
```
Comment (sebastiandedeyne) [inline src/MailableFactory.php]: Is this necessary? We're already doing a `class_exists` in `SendTestMail`, can't we normalize the class name there so this class doesn't care about resolving the base namespace?

**Review by sebastiandedeyne — CHANGES_REQUESTED** [review]
Big thanks for this! I posted a few more change requests, when these are done I'll pull this in!

**Comment by fetzi** [comment]
@sebastiandedeyne done, I've also removed the `MailableDoesNotExist` exception class.

**Comment by sebastiandedeyne** [comment]
Thanks! Released as `v1.1`

**Pull request merged**